### PR TITLE
Add JSitor liquid tag

### LINKED ISSUE #4214: Request to support JSitor Embeddable
Hi Dev.to Team,

This is a feature request to support JSitor as embeddable content on Dev.to article sites. I have seen that JSFiddle and Codepen are supported in Dev.to which is great and same way we would like to add the support for JSitor as well.

The pattern followed for Codepen is as such
{% codepen https://codepen.io/twhite96/pen/XKqrJX %}

We would like to use the same pattern. JSitor already supports oembed protocol if that is the integration required. If not, the main site generally is loaded with below URL
https://jsitor.com/IfRnIEscb

and embeddable formats can be loaded with 
https://jsitor.com/embed/IfRnIEscb

So I would like to have this support as such
{% jsitor https://jsitor.com/embed/IfRnIEscb %}

### PR DESCRIPTION
**Pull request by artoodeeto**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Documentation Update

## Description
Added a feature for jsitor liquid

## Related Tickets & Documents
Resolves #4214

## Added to documentation?

- [ ] docs.dev.to
- [ ] readme
- [ ] no documentation needed

## [optional] What gif best describes this PR or how it makes you feel?

alt_text

### CODE REVIEW

**Comment by artoodeeto** [comment]
I didn't include any images because its already in #4214 but if you want ill add it. Please let me know where can I improve more and also sorry for the delay had to bring home the work. 🙏

File: app/liquid_tags/jsitor_tag.rb [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+    jsitor_error
+  end
+
+  def link_valid?(link)
+    JSITOR_URL.match link
+  end
+
+  def jsitor_error
+    raise StandardError, "Invalid jsitor link"
```
Comment (ashvin777) [inline app/liquid_tags/jsitor_tag.rb]: Suggestion - If we can add specific error by adding the example link then it will be more awesome. :-)

File: app/liquid_tags/jsitor_tag.rb [hunk 2]
```diff
@@ -0,0 +1,38 @@
+class JsitorTag < LiquidTagBase
+  PARTIAL = "liquids/jsitor".freeze
+  JSITOR_URL = /\A(https|http):\/\/(jsitor)\.(com)\/(embed)\/([a-zA-Z0-9]+)\Z/.freeze
+  def initialize(tag_name, link, token)
```
Comment (rhymes) [inline app/liquid_tags/jsitor_tag.rb]: you could add a space here

File: app/liquid_tags/jsitor_tag.rb [hunk 3]
```diff
@@ -0,0 +1,38 @@
+class JsitorTag < LiquidTagBase
+  PARTIAL = "liquids/jsitor".freeze
+  JSITOR_URL = /\A(https|http):\/\/(jsitor)\.(com)\/(embed)\/([a-zA-Z0-9]+)\Z/.freeze
```
Comment (rhymes) [inline app/liquid_tags/jsitor_tag.rb]: `URL_REGEXP` seems a better name :)

File: app/liquid_tags/jsitor_tag.rb (hunk 1 above)
Comment (rhymes) [reply app/liquid_tags/jsitor_tag.rb]: also `JSitor` instead of `jsitor` :)

File: app/views/pages/_editor_guide_text.html.erb [hunk 4]
```diff
@@ -328,6 +328,9 @@
         <h3><strong>Asciinema Embed</strong></h3>
         <p>All you need is the Asciinema id:</p>
         <code>{% asciinema 239367 %}</code>
+        <h3><strong>Jsitor Liquid Tag</strong></h3>
```
Comment (rhymes) [inline app/views/pages/_editor_guide_text.html.erb]: `JSitor`

File: app/views/pages/_editor_guide_text.html.erb [hunk 5]
```diff
@@ -328,6 +328,9 @@
...
+        <p>Copy jsitor embeddable link make sure no slash (/) on the end of the <code>link</code>, it should have the <em>embed</em> path, and ends with ID code, as follows:</p>
```
Comment (rhymes) [inline app/views/pages/_editor_guide_text.html.erb]: `JSitor`

**Comment by artoodeeto** [comment]
@ashvin777 @rhymes hello guys. thanks for the comments. I will add theses changes tonight. also my bad for the branch name. thought it was alright. :) thanks again

**Comment by artoodeeto** [comment]
Jesus Christ Sorry I made a mistake on rebasing. 😢

**Comment by artoodeeto** [comment]
@rhymes sir I messed up. I updated my origin branch. what I did was on my master branch:

1. git pull upstream master
2. git rebase upstream/master
3. git push origin master
4. checked out this feature branch
5. git rebase master
6. git push origin/feature-branch

and then there were commits that unrelated to this branch so I googled and followed [this](https://stackoverflow.com/a/51400593/7034540). 

basically, to fix it I dropped the 23 unrelated commits:
1. git rebase -i HEADE~23
2. drop the 23
3. save and exit
4. git push --force-with-lease

but I dont know if it will affect the history. im really sorry. also on my last commit there's a big green maybe thats on the space between regex and initialize method? my bad. sorry. next time ill be very careful and if I get ban I understand.

**Comment by rhymes** [comment]
@artoodeeto don't worry, git mistakes happen! the great thing is that they are usually fixable. i'll get another review in

**Comment by artoodeeto** [comment]
@rhymes I'm not really sure how to test it now. removed a whole lot since I just need to accept the id then recreate the link. I don't have really to check the id that is being passed because it will be parse as a string anyway. what do you think?

File: app/views/liquids/_jsitor.html.erb [hunk 6]
```diff
@@ -0,0 +1,8 @@
+<iframe
+  height="<%= height %>"
+  src="<%= link %>"
+  scrolling="no"
+  frameborder="no"
+  allowtransparency="true"
+  style="width: 100%;">
```
Comment (rhymes) [inline app/views/liquids/_jsitor.html.erb]: could you add `loading="lazy"` to this? Since there's a PR going around adding lazy loading to iframes, we might as well get that in right away here :D

Comment (artoodeeto) [reply app/views/liquids/_jsitor.html.erb]: ohh. hey sir is that the right behavior? @ashvin777. 

@rhymes ill add it when I get back home :smile: also are my test suite fine? I dont know how to test that kind of behavior, but in Integration test maybe?

Comment (rhymes) [reply app/views/liquids/_jsitor.html.erb]: @artoodeeto I think we should be fine without testing the "defaultness" of jsitor when the id is incorrect

Comment (ashvin777) [reply app/views/liquids/_jsitor.html.erb]: Thank @rhymes for noticing this issue. I will fix this behavior. Ideally, it should show a "404" error page.

**Comment by ashvin777** [comment]
Hi @rhymes and @artoodeeto , is it possible to convert the regular link `https://jsitor.com/B7FQ5tHbY` into embed link `https://jsitor.com/embed/B7FQ5tHbY`  internally ?

So if user enters the embed template like below 
`${ jsitor https://jsitor.com/B7FQ5tHbY }`

It should load `https://jsitor.com/embed/B7FQ5tHbY` link instead?

File: app/views/liquids/_jsitor.html.erb (hunk 6 above)
Comment (rhymes) [reply app/views/liquids/_jsitor.html.erb]: I think the current behavior is fine as well, I'll let you evaulate how hard it'd be to intercept something like that

**Comment by rhymes** [comment]
@ashvin777 good idea

File: app/views/liquids/_jsitor.html.erb (hunk 6 above)
Comment (artoodeeto) [reply app/views/liquids/_jsitor.html.erb]: @ashvin777 @rhymes we could check the response status? make an HTTP request on the backend check status if 2** then continue loading the page 4** then raise an error? what do you think?

**Comment by artoodeeto** [comment]
> Hi @rhymes and @artoodeeto , is it possible to convert the regular link `https://jsitor.com/B7FQ5tHbY` into embed link `https://jsitor.com/embed/B7FQ5tHbY` internally ?
> 
> So if user enters the embed template like below
> `${ jsitor https://jsitor.com/B7FQ5tHbY }`
> 
> It should load `https://jsitor.com/embed/B7FQ5tHbY` link instead?

@rhymes @ashvin777  but we already have the guide though? if they check the guide they will read that it's just the id is being required to load it. no?

File: app/views/liquids/_jsitor.html.erb (hunk 6 above)
Comment (rhymes) [reply app/views/liquids/_jsitor.html.erb]: Don't think so.

```shell
➜  ~ curl -I https://jsitor.com/foobar
HTTP/1.1 200 OK
➜  ~ curl -I https://jsitor.com/embed/foobar
HTTP/1.1 200 OK
```

I'm thinking we shoulnd't get stuck on this detail, it likely won't matter much because the publisher will likely check if their embed is correct before publishing.

As long as there are no security fallacies here I don't think spending time to intercept if the ID is correct or not is worth it.

Comment (artoodeeto) [reply app/views/liquids/_jsitor.html.erb]: @rhymes  I agree with you. also we have the guide to "guide" them lols. but we could technically do it if @ashvin777 will update the site to return 400.

**Comment by rhymes** [comment]
@artoodeeto yeah, I think it can be as is right now, but there are liquid tags like the listings tag that supports both the "id" and the URL :)

If you add the lazy loading bit I think the PR can be accepted as is, bar a second set of eyes in case I've missed something

File: app/views/liquids/_jsitor.html.erb (hunk 6 above)
Comment (artoodeeto) [reply app/views/liquids/_jsitor.html.erb]: heres my fix for lazy loadig. 😄 @rhymes

**Comment by artoodeeto** [comment]
@ashvin777 @rhymes welcome welcome. :smile:

**Comment by ashvin777** [comment]
Hi @artoodeeto , Thanks you so much for this work. 

I just saw the production guide (dev.to) of JSitor published after merging this PR

image

Which seems incorrect, The request was to use the URL instead of the id. The reason behind that is, there are some additional query params which can be used to hide/show some of the tabs in embeddable content, for example, to see the JS tab content only then we can do it using https://jsitor.com/embed/1QgJVmCam?js

A light theme embeddable -
https://jsitor.com/embed/1QgJVmCam?html&js&css&result&light&

All of these options are configurable in embed section of JSitor
image

Is it possible to either change this to use URL or else is there any other way of passing this query params from this liquid tag?

**Comment by Zhao-Andy** [comment]
Hi @ashvin777, thanks for raising that. I missed your comment earlier about it before merging it. I think having the current iteration is a good start, and we can add the ability to handle query parameters as well as use the full URL easily.

@artoodeeto if you'd like to continue with this and work on that, you're more than welcome to! Otherwise I'll make an issue and leave the task up for grabs.

**Comment by artoodeeto** [comment]
@ashvin777 sorry about that, I thought we only need the embed with id so I didn't bother about the query params. ill redo this. :smile: probably this weekend.

@Zhao-Andy can you make an issue on this one and assign it to me? I don't want to use the current branch because it's broken. If I make a push its always asking me to pull first so I have to do a force push. if it's ok with you. :)

@rhymes hello sir just wanted to tag you. haha. :v:

**Comment by rhymes** [comment]
@artoodeeto yeah, you shouldn't use a merged branch, you can delete it and branch off from a fresh master, don't forget to update your copy of master with `git checkout master && git pull upstream master`

**Comment by artoodeeto** [comment]
@rhymes yeap. Ill im gonna grab the new issue. thank you! :heart:
